internal/container/reference.go

```diff
@@ -67,11 +68,14 @@ func (rs RefSet) WithoutRegistry() RefSet {
 	return MustSimpleRefSet(rs.ConfigurationRef)
 }
 
-func (rs RefSet) Registry() Registry {
-	return rs.registry
+func (rs RefSet) Registry() *v1alpha1.RegistryHosting {
+	if rs.registry == nil {
+		return nil
+	}
+	return rs.registry.DeepCopy()
```

Comment:
TBH i'm surprised we're doing a copy so often. was mutating registries a problem, or is this just being defensive?

(one of my big complains about go is the lack of mutability protection, but i do like the Kubernetes convention of "always copy before mutating")

Reply:
Well, previously it was always used as a value type in conjunction with the `Registry::Empty()` method, so we weren't handing out mutable references here

FWIW, in cases like this, `DeepCopy()` allocates and then does `*out = *in`, so minus the unavoidable heap usage, it shouldn't be worse